persistence-api/src/main/java/io/stargate/db/cdc/CDCServiceImpl.java

```diff
@@ -0,0 +1,142 @@
+          1, new ThreadFactoryBuilder().setDaemon(true).setNameFormat("failAfter-%d").build());
+
+  static {
+    unhealthyFuture.completeExceptionally(
+        new CDCWriteException("CDC producer marked as unhealthy"));
+  }
+
+  public CDCServiceImpl(CDCProducer producer, MetricRegistry registry) {
+    // TODO: Wrap configuration and use default health checker
+    this(producer, null, null, registry);
+  }
+
+  @VisibleForTesting
+  public CDCServiceImpl(
+      CDCProducer producer,
+      CDCHealthChecker healthChecker,
+      CDCConfig config,
+      MetricRegistry registry) {
+    this.producer = producer;
+    this.healthChecker = healthChecker;
+    this.config = config;
+
+    CDCMetrics.instance.init(registry);
+  }
+
+  @Override
+  public CompletableFuture<Void> publish(MutationEvent mutation) throws CDCWriteException {
+    if (!config.isTrackedByCDC(mutation)) {
+      return completedFuture;
+    }
+
+    if (!healthChecker.isHealthy()) {
+      return unhealthyFuture;
+    }
+
+    return sendToProducer(mutation);
+  }
+
+  private CompletableFuture<Void> sendToProducer(MutationEvent mutation) {
+    final long start = System.nanoTime();
+    CDCMetrics.instance.incrementInFlight();
+    return orTimeout(producer.publish(mutation), config.getLatencyErrorMs())
+        .whenComplete(
+            (r, e) -> {
+              CDCMetrics.instance.decrementInFlight();
+              CDCMetrics.instance.updateLatency(System.nanoTime() - start);
+
+              if (e != null) {
+                healthChecker.reportSendError(mutation, e);
+                CDCMetrics.instance.markProducerFailure();
+
+                if (e instanceof TimeoutException) {
+                  CDCMetrics.instance.markProducerTimedOut();
+                }
+              } else {
+                healthChecker.reportSendSuccess();
+              }
+            });
+  }
+
+  /** Equivalent of Java 9+ orTimeout() */
+  private <T> CompletableFuture<T> orTimeout(CompletableFuture<T> f, long timeoutMs) {
+    if (timeoutMs <= 0) {
+      return f;
+    }
+
+    try {
+      final ScheduledFuture<Boolean> scheduledTimeout =
+          timeoutScheduler.schedule(
+              () ->
+                  f.completeExceptionally(
+                      new TimeoutException(String.format("Timed out after %sms", timeoutMs))),
+              timeoutMs,
+              MILLISECONDS);
+
+      f.thenAccept(r -> scheduledTimeout.cancel(false));
+    } catch (Exception ex) {
+      // Scheduler was shutdown, nvm
+    }
+
+    return f;
+  }
+
+  @Override
+  public void close() throws Exception {
+    healthChecker.close();
+    timeoutScheduler.shutdownNow();
+    try {
+      producer.close().get();
+    } catch (Exception e) {
+      logger.info("There was an issue releasing resources of CDC Producer", e);
+    }
+  }
+
+  @VisibleForTesting
+  interface CDCConfig {
+    boolean isTrackedByCDC(MutationEvent mutation);
+
+    long getLatencyErrorMs();
```

Comment:
nit, the name of this method: `getLatencyErrorMs`  does not seem clear to me. Can we change it or add docs?

Reply:
I've changed it to a more clear name.